### ISSUE TITLE
FR: Support NetworkManager and Firewalld

### ISSUE THREAD

**Issue by ananthb** [tailscale/tailscale#3416]
### What are you trying to do?

I would like to assign a firewalld firewall zone to tailscale0 interface on Linux permanently.

### How should we solve this?

Detect presence of NetworkManager and Firewalld and create a NetworkManager configuration for tailscale0 that contains a user selected zone. The zone to be used for tailscale0 could be exposed as a preference to users. Alternatively, network manager dispatcher can be used to perform this action on startup every time tailscale0 comes up.

### What is the impact of not solving this?

On every boot, I manually assign tailscale0 to a zone.

### Anything else?

https://wiki.archlinux.org/title/Firewalld#Using_NetworkManager_to_manage_zones

https://wiki.archlinux.org/title/NetworkManager#Network_services_with_NetworkManager_dispatcher

**Comment by danderson** [comment]
In general, we can't safely use NetworkManager in tailscale, because once we ask it to do anything to tailscale0, it wants to own the entire interface configuration, and that ends up with a fractal of bugs and different behaviors across different versions of NetworkManager.

I don't know firewalld, does it expose a DBus interface we can talk to directly, similarly to systemd-resolved? That would be tractable, and would make this work even on servers that run firewalld but not NM.

**Comment by ananthb** [comment]
Firewalld does expose a [DBus interface](https://firewalld.org/documentation/man-pages/firewalld.dbus.html).

**Comment by lordraiden** [comment]
If don't use firewalld or if I use ufw, do I have to change something?

**Comment by ananthb** [comment]
Tailscale creates the `tailscale0` interface which can be configured using any firewall application. This issue is about automated configuration with firewalld zones. You can configure ufw for `tailscale0` just like you would for any other network interface on your machine.

**Comment by lordraiden** [comment]
> Tailscale creates the `tailscale0` interface which can be configured using any firewall application. This issue is about automated configuration with firewalld zones. You can configure ufw for `tailscale0` just like you would for any other network interface on your machine.

So regarding this for oracle linux
![imagen](https://user-images.githubusercontent.com/17175395/165584048-0cfd0b36-a0c9-4451-b2d7-3e587fdb8248.png)
and this for centos 8
![imagen](https://user-images.githubusercontent.com/17175395/165584113-14177fb1-f5c8-4ad9-81d5-8ef03ab0b163.png)

Is the problem and the solution the same for both? 
If I am going to use subnet routers in Centos can I apply this workaround? "firewall-cmd --permanent --add-masquerade" or it's simply not compatible as stated above?

Both links in the documentation point to this github issue

On the other hand what repo should we use with distros like almalinux? oracle or centos?

**Comment by DentonGentry** [comment]
@lordraiden if you want to bring up topics in distributions which happen to use firewalld please either open new bugs or email support@tailscale.com.

**Comment by twouters** [comment]
Workaround: edit `/etc/firewalld/zones/<zone>.xml` and add an [interface element tag](https://firewalld.org/documentation/man-pages/firewalld.zone.html): `<interface name="tailscale0"/>`.

**Comment by ananthb** [comment]
@twouters that worked, thanks!

**Comment by twouters** [comment]
For the record, I had to tell NetworkManager to not manage the connection (I think it changes the connection zone)

Create file `/etc/NetworkManager/conf.d/unmanage-tailscale.conf`

```
# do not intervene with firewalld zoning
[keyfile]
unmanaged-devices=interface-name:tailscale0
```

It might also be possible to define the zone through NetworkManager instead of firewalld.

**Comment by ananthb** [comment]
Huh. It worked for me with just the edit to the zone.xml file. 🤷

**Comment by twouters** [comment]
Yeah, it worked for me too when I mentioned the workaround but after a while (not sure what caused it) it got assigned to the public zone.  
I'll try to figure out the cause when I find the time.

**Comment by raggi** [comment]
Recent work on #9084 and #391 may help here, but I have not yet investigated the outcome on a firewalld system.

**Comment by soulflyman** [comment]
> # do not intervene with firewalld zoning
> [keyfile]
> unmanaged-devices=interface-name:tailscale0

I also had to add this config to `/etc/NetworkManager/conf.d/unmanage-tailscale.conf` in addition to the firewalld command that adds the home zone permanently

**Comment by baggiiiie** [comment]
got directed here by [tailscale's official doc on setting up subnet router](https://tailscale.com/kb/1019/subnets#enable-ip-forwarding), im wondering how does this issue affect subnet router, and how does allowing ip masquerading on firewalld help with this issue?

will be appreciated if someone in the know can help out!

**Comment by xpseudonym** [comment]
I think this bug should be closed - reading through the posts, it feels like it's taken 5 years to read the firewalld manual...:
[firewalld: A service daemon with D-Bus interface](https://firewalld.org/)
[A beginner's guide to firewalld in Linux](https://www.redhat.com/en/blog/beginners-guide-firewalld)
[firewall-cmd](https://firewalld.org/documentation/man-pages/firewall-cmd.html)

Here's a straight forward example of how to configure firewalld for Tailscale, [Secure Tailscale networks with firewalld](https://major.io/p/secure-tailscale-networks-with-firewalld/)
